### DISCUSSION TITLE
The library could not display some color correctly

### DISCUSSION THREAD

**Discussion by appleshaman** [Bodmer/TFT_eSPI#2322] [Q&A - General] [answered]
I tried to fill the colour with 0x3D8B, which is a green colour, I used both pushImage(0x3D8B) and fillScreen(), but both of them returned the wrong colour, which is close to blue colour. The only thing that can display the colour correctly is fillScreen(tft.color565(56, 178, 92)), which is I have to use color565 to let the colour display correctly. but if I want to show some pictures with pushImage(), then I could not display the image's colour correctly. 
I also tried some predefined colours like TFT_BLUE stored in the library and they works fine
The IDE is PlatformIO
Library version
Board I chose ttgo-lora32-v1 in PlatformIO but the hardware itself was ttgo T-display as there is no such board in IDE
Processor is esp32
TFT driver: not sure but I think ttgo t-display use this one: http://www.lcdwiki.com/1.14inch_IPS_Module

the color 0x3D8B it displayed is like this:
![9](https://user-images.githubusercontent.com/76955501/212469815-87186061-97a1-4775-a01f-ee36298d584d.jpg)

[main.zip](https://github.com/Bodmer/TFT_eSPI/files/10417218/main.zip)

**Comment by Bodmer** [comment, marked as answer]
The screen has a 16 bit colour interface  this is the 565 RGB format. So 24 bit 888 format colours must be converted to a new 16 bit value. That is what the color565 function does. If you try to write:
tft.fillScreen(0x38b25c)

Then only the 16 bits 0xb25c will be used.

**Reply by appleshaman** [reply]
Hello Bodmer
I appreciate your help, Now I got what is happening. However, I still have one question, that is when I tried to convert some images to C file, it is like it used a different colour format; like what I converted in the C file is 0x3D8B but what the screen displayed by using pushImage() is more like the colour 0x003D8B in 24 bit, is that because there is some more different format?
What converter I used is here: http://www.rinkydinkelectronics.com/t_imageconverter565.php
it is like 565 format but still could not display correctly.

**Reply by appleshaman** [reply]
> Hello Bodmer I appreciate your help, Now I got what is happening. However, I still have one question, that is when I tried to convert some images to C file, it is like it used a different colour format; like what I converted in the C file is 0x3D8B but what the screen displayed by using pushImage() is more like the colour 0x003D8B in 24 bit, is that because there is some more different format? What converter I used is here: http://www.rinkydinkelectronics.com/t_imageconverter565.php it is like 565 format but still could not display correctly.

Sorry, Forget about it, I just realized that I have to use setSwapBytes() before I print a image

**Comment by nodemand** [comment]
Hi!
I can not get the colors right, even with the suggestions above. The screen should display blue rectangles, but I get yellows...
I'm using library version 2.5.34 from the Arduino Library manager with a WaveShare Pico-ResTouch-LCD-3.5 display. 
Code:
```
  tft.fillScreen(tft.color565(0xff,0xff,0xff));
  tft.fillRect(0, 0, 320, 240, tft.color565(0,0,0x88));
  tft.fillRect(0, 240, 160, 480, tft.color565(0,0,0x44));
  tft.fillRect(160, 240, 320, 480, tft.color565(0,0,0x66));
```

Shows:
![IMG_6170](https://github.com/Bodmer/TFT_eSPI/assets/8333808/6f41260c-c475-4999-82ac-3b6554f72db2)

What am I doing wrong? I tried tft.setSwapBytes(true) and tft.setSwapBytes(false). Also the calibration screen is off. It should be red and white, but it is black and blue...

`tft.calibrateTouch(calibrationData, TFT_WHITE, TFT_RED, 15);`

Shows:
![IMG_6169](https://github.com/Bodmer/TFT_eSPI/assets/8333808/d3f3cc78-bb89-4b9c-86cf-a6c7377e7acf)

Help!

**Reply by alexandrglm** [reply]
> Hi! I can not get the colors right, even with the suggestions above. The screen should display blue rectangles, but I get yellows... I'm using library version 2.5.34 from the Arduino Library manager with a WaveShare Pico-ResTouch-LCD-3.5 display. Code:
> 
> ```
>   tft.fillScreen(tft.color565(0xff,0xff,0xff));
>   tft.fillRect(0, 0, 320, 240, tft.color565(0,0,0x88));
>   tft.fillRect(0, 240, 160, 480, tft.color565(0,0,0x44));
>   tft.fillRect(160, 240, 320, 480, tft.color565(0,0,0x66));
> ```
> 
> Shows: ![IMG_6170](https://private-user-images.githubusercontent.com/8333808/294021370-6f41260c-c475-4999-82ac-3b6554f72db2.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc3Nzc0OTIsIm5iZiI6MTc0Nzc3NzE5MiwicGF0aCI6Ii84MzMzODA4LzI5NDAyMTM3MC02ZjQxMjYwYy1jNDc1LTQ5OTktODJhYy0zYjY1NTRmNzJkYjIuanBlZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTIwVDIxMzk1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFkZmRlNzM1NDQxZmY1YmQ5Y2RkNDg5NzlkY2ZjNDdiYmJiMjYyY2MxOTkzOTI4N2Y3YzhlNzg3NTAzYmQ4MDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.7dXflLrVQFBB1gpsID-ZY4ZNxloHnV8fAyyHbDFpgyE)
> 
> What am I doing wrong? I tried tft.setSwapBytes(true) and tft.setSwapBytes(false). Also the calibration screen is off. It should be red and white, but it is black and blue...
> 
> `tft.calibrateTouch(calibrationData, TFT_WHITE, TFT_RED, 15);`
> 
> Shows: ![IMG_6169](https://private-user-images.githubusercontent.com/8333808/294021706-d3f3cc78-bb89-4b9c-86cf-a6c7377e7acf.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc3Nzc0OTIsIm5iZiI6MTc0Nzc3NzE5MiwicGF0aCI6Ii84MzMzODA4LzI5NDAyMTcwNi1kM2YzY2M3OC1iYjg5LTRiOWMtODZjZi1hNmM3Mzc3ZTdhY2YuanBlZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTIwVDIxMzk1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM5MWU5NmVjZWQ4Mjg2NjYzZGQ1ZDRhZTRkMDc2ZGQ3ZTE5NWM5ZWMyY2U1YWIyNTgzM2MwZDlmMzNlOTJkMWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.7Tq7cFo4r8DhXBYcB1jjFcMnSc1CLJU-vz3BpNyFPg8)
> 
> Help!

Firstly, forgive me for bumping in a previous answered thread but, sir, I'm using a ESP32 devmodule-alike (ESP32-2432S028R board with a similar TFT as your module is) and I can't get colours in a proper way even using _TFT_<COLOUR>_ vars. (TFT_WHITE, TFT_BLACK, TFT_RED, etc).  

Blue is yellow, yellow is purple, black is white, ...  The colours are ... upside down to each contrast colour!  

I think it's not a problem with the TFT_eSPI component, but a problem with the spi - gpio connection instead.  

If you found a solution and want to share here, I'll appreciate.

Greetings.
Alexandr.

**Reply by nodemand** [reply]
Try this:

tft.invertDisplay(1);

**Reply by alexandrglm** [reply]
> Try this:
> 
> tft.invertDisplay(1);

Thanks sir, it works!
I'd like to understand if this is a common issue in any board, or it's related to several vendors ... maybe a wrong hi/low pin conf ... spi conf... or ... why!

Anyway, again, thanks!!!!
